[Feat] Allow more flexible import paths of `CapturedCallable` (plus bypassing validation) (#1255)
mckinsey/vizro: vizro-core/docs/pages/user-guides/extensions.md

It is possible to create custom React.js components and add them directly to any Vizro dashboard so enabling you to code beneath both the Vizro and Dash layers and control React.js directly

For more information, see the documentation on [using React.js components with Dash](https://dash.plotly.com/plugins)

> ## Using custom functions in `yaml`/`json` configuration

Reviewer: I think section should go as a subsection of the `Vizro customizations` section.

> ## Using custom functions in `yaml`/`json` configuration

Reviewer: ```suggestion
## Use custom functions in `yaml`/`json` configuration
```

!!! note "Exposing configuration to untrusted users"

>     Exposing Vizro configuration to untrusted users may pose a security risk. By choosing a specific import path for a custom function, it could be possible to inject malicious code that will be executed when the dashboard is run. Note that for this to work, the malicious code would need to be executed on import, and it would need to be part of a dependency that is available at validation time.

Reviewer: ```suggestion
    Exposing Vizro configuration to untrusted users may pose a security risk. A user with access to YAML/JSON configuration can potentially execute arbitrary Python code when the dashboard is run. The security of your `yaml`/`json` configuration should be regarded as equivalent to the security of your `app.py` file.
```

> It is possible to refer to custom functions that are used as `CapturedCallable` by their import path in a `yaml`/`json` configuration of the dashboard.
> 
> In the [documentation sections on custom charts, tables, figures and actions](#vizro-customizations), you will find examples on how the Vizro schema can be extended by using custom Python code. It is possible to refer to these custom functions in the `yaml`/`json` configuration by using the `_target_` key and the correct import path.

Reviewer: ```suggestion
It is possible to refer to custom functions for charts, tables, figures and actions by their import path in a `yaml`/`json` configuration of the dashboard. In the documentation on custom charts, tables, figures and actions, you will find examples of how the Vizro schema can be extended by using custom Python code. It is possible to refer to these custom functions in the `yaml`/`json` configuration by using the `_target_` key and the correct import path.
```

> !!! example "Custom charts with YAML config example"

Reviewer: Please can we do this as multiple tabs that show all the files required rather than embedding a YAML string inside the Python code? I'd suggest:
* one file with our standard `app.py` for loading YAML 
* one file `custom_charts.py` with `bar` defined
* one `dashboard.yaml` file with `_target_: custom_charts.bar`

Note the yaml file shouldn't need any `"` marks anywhere I believe.

Then afterwards in a note or in a  `(1)!` thing like you do now, let's say that you could define the chart in `app.py` itself and refer to it as `__main__.bar`.

This way we're demonstrating the "normal" looking `custom_charts.bar` first and then the weird looking `__main__` afterwards.

Also would be worth linking to pydantic's docs on `ImportString` that explain the format of string expected.

Author: Ok will do although I did this on purpose because I like example that are simply copy past and run as is.

    def custom_bar(data_frame: pd.DataFrame, x: str, y: str) -> go.Figure: # (1)!
        """Custom bar chart using Plotly Graph Objects."""
        return go.Figure(data=[go.Bar(x=data_frame[x], y=data_frame[y])])

    # Dashboard configuration in YAML format
    dashboard_yaml = """
    title: "Custom Chart Example"
    pages:
        - title: "Custom Bar Chart"
            components:
                - type: "graph"
                    figure:
                        _target_: "__main__.custom_bar"
                        data_frame: "iris"
                        x: "sepal_length"
                        y: "sepal_width"
    """

    # Load dashboard from YAML
    dashboard_config = yaml.safe_load(dashboard_yaml)# (2)!
    dashboard = vm.Dashboard(**dashboard_config)

    # Build and run the dashboard
    app = Vizro().build(dashboard)

    if __name__ == "__main__":
        app.run()
    ```

    1. Define the custom chart either in `app.py` or in a separate file.
    1. Parse the YAML or JSON configuration that refers to the custom chart with the **correct import path**. If the custom chart is defined in the `app.py`, then use `__main__` as the import path.

> ### Validating dashboards without executing `CapturedCallable` functions

Reviewer: ```suggestion
### Validate dashboards without defining `CapturedCallable` functions
```

> It is possible to validate a dashboard configuration without executing the `CapturedCallable` functions.

Reviewer: ```suggestion
It is possible to validate a dashboard configuration without importing or executing some `CapturedCallable` functions.
```

You can use this method when you want to check if the dashboard configuration is valid, but you don't want to execute the custom functions until run-time, which may be in a sandboxed environment. This is useful when the custom functions are not available at validation time, or when they originate from untrusted sources (e.g. when a large language model is used to generate that code).

!!! note When starting the server

>     If you validate your dashboard configuration with a `CapturedCallable` function that is undefined, you will not be able to start the server. Hence when running the dashboard, you will need to recreate a dashboard object with all the `CapturedCallable` functions defined.

Author: Yes, that is a very good idea. Let me add a TODO to the code

!!! example "Validating dashboards without executing `CapturedCallable` functions"

    ```python
    import vizro.models as vm
    import vizro.plotly.express as px
    from vizro import Vizro
    from vizro.managers import data_manager
    from vizro.models.types import AllowedCapturedCallable

    data_manager["iris"] = px.data.iris()

    dashboard_config = {
        "title": "Test dashboard",
        "pages": [
            {
                "title": "Page 1",
                "components": [
                    {
                        "type": "ag_grid",
                        "figure": {"_target_": "llm_generated_grid", "data_frame": "iris"},
                    },
                ],
            }
        ],
    }

    dashboard = vm.Dashboard.model_validate(
        dashboard_config,
        context={
            "allowed_undefined_captured_callables": [
                AllowedCapturedCallable(function_name="llm_generated_grid", mode="ag_grid")
            ]
        },
    )

>     app = Vizro().build(dashboard)  # (1)!
>     ```
> 
>     1. Because this dashboard configuration contains a `CapturedCallable` function that is undefined, you will not be able to start the server. Hence `app.run()` will raise an error.

Reviewer: Something like this might be clearer? Assuming I understood the behaviour correctly.

```suggestion
    app = Vizro().build(dashboard)  # (1)!
    try:
        app.run()  # (2)!
    except Exception as exc:
        print(exc)
    ```

    1. The dashboard configuration contains a `CapturedCallable` function that is undefined but allowed by `allowed_undefined_captured_callables`. The app can still be built without raising any errors.
    2. However, it is not possible to run the app without undefined `CapturedCallable`s. This raises an error.
```
